### ISSUE TITLE
3D convolutions on GPU with large input produces incorrect results on some GPUs

### ISSUE THREAD

**Issue by gerbenvv** [tensorflow/tensorflow#22095]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes, see reproduction script below.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04.5 LTS
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: N/A
- **TensorFlow installed from (source or binary)**: Binary via pip
- **TensorFlow version (use command below)**: 1.10.1
- **Python version**: 3.6.3
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: GCC 5.4.0
- **CUDA/cuDNN version**: CUDA 9.0 / cuDNN 7.2
- **GPU model and memory**: GTX 1080 Ti
- **Exact command to reproduce**: See script below.

```
== cat /etc/issue ===============================================
Linux gpu01 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
VERSION="16.04.5 LTS (Xenial Xerus)"
VERSION_ID="16.04"
VERSION_CODENAME=xenial

== are we in docker =============================================
No

== compiler =====================================================
c++ (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


== uname -a =====================================================
Linux gpu01 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy                    1.14.2
protobuf                 3.6.1
tensorflow-gpu           1.10.1
tensorflow-tensorboard   1.5.1

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.10.1
tf.GIT_VERSION = v1.10.1-0-g4dcfddc5d1
tf.COMPILER_VERSION = v1.10.1-0-g4dcfddc5d1
Sanity check: array([1], dtype=int32)

== env ==========================================================
LD_LIBRARY_PATH /usr/local/cuda/lib64
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
Wed Sep  5 16:19:24 2018
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 387.26                 Driver Version: 387.26                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 108...  On   | 00000000:04:00.0 Off |                  N/A |
| 23%   38C    P0    64W / 250W |      1MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 108...  On   | 00000000:05:00.0 Off |                  N/A |
| 23%   36C    P2    65W / 250W |      1MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce GTX 108...  On   | 00000000:08:00.0 Off |                  N/A |
| 23%   42C    P0    77W / 250W |      1MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  GeForce GTX 108...  On   | 00000000:09:00.0 Off |                  N/A |
| 23%   42C    P0    63W / 250W |      1MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   4  GeForce GTX 108...  On   | 00000000:85:00.0 Off |                  N/A |
| 23%   35C    P0    63W / 250W |      1MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   5  GeForce GTX 108...  On   | 00000000:86:00.0 Off |                  N/A |
| 23%   38C    P0    73W / 250W |      1MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   6  GeForce GTX 108...  On   | 00000000:89:00.0 Off |                  N/A |
| 23%   38C    P0    72W / 250W |      1MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   7  GeForce GTX 108...  On   | 00000000:8A:00.0 Off |                  N/A |
| 23%   36C    P0    71W / 250W |      1MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/cuda-8.0/doc/man/man7/libcudart.so.7
/usr/local/cuda-8.0/doc/man/man7/libcudart.7
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudart.so.8.0.61
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudart_static.a
/usr/local/cuda-9.0/doc/man/man7/libcudart.so.7
/usr/local/cuda-9.0/doc/man/man7/libcudart.7
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudart.so.9.0.176
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudart_static.a
/usr/local/cuda-9.1/doc/man/man7/libcudart.so.7
/usr/local/cuda-9.1/doc/man/man7/libcudart.7
/usr/local/cuda-9.1/targets/x86_64-linux/lib/libcudart.so.9.1.85
/usr/local/cuda-9.1/targets/x86_64-linux/lib/libcudart_static.a
```

### Describe the problem
3D convolutions on GPU with large batches fail to calculate the correct result for the last part of the batch. The first entries will be correct, however, after some point the resulting values will be random. This is the case for at least the GTX 1080 Ti and the GeForce GTX 980 Ti gpus. It produces the correct result however on the CPU and at least the K80 GPU.

### Source code / logs

A small reproduction script:

```
import tensorflow as tf
import numpy as np



run_on_gpu = True  # Set to false for the correct result, computed on the CPU.
batch_size = 128  # Try out a large batch, that just fits into memory. Small batches work fine.



random_state = np.random.RandomState(42)

inputs = tf.placeholder(tf.float32, (None, 29, 72, 72, 1), 'inputs')

device_name = '/gpu:0' if run_on_gpu else '/cpu:0'

with tf.device(device_name):
    weights = tf.constant(random_state.randn(3, 3, 3, 1, 32).astype(np.float32))
    outputs = tf.nn.conv3d(inputs, weights, strides=(1, 1, 1, 1, 1), padding='SAME', dilations=(1, 1, 1, 1, 1))
    weights = tf.constant(random_state.randn(1, 1, 1, 32, 16).astype(np.float32))
    outputs = tf.nn.conv3d(outputs, weights, strides=(1, 1, 1, 1, 1), padding='VALID', dilations=(1, 1, 1, 1, 1))

session = tf.Session()

random_inputs = random_state.randn(batch_size, 29, 72, 72, 1).astype(np.float32)

a = session.run(outputs, feed_dict={inputs: random_inputs})[-1]
b = session.run(outputs, feed_dict={inputs: random_inputs[-1:]})[0]

# Should print the same line, twice.
print(a.shape, a.ravel()[0])
print(b.shape, b.ravel()[0])
```

This script should print the same line twice.

Expected result:
```
(29, 72, 72, 16) -5.843975
(29, 72, 72, 16) -5.843975
```

Actual result:
```
(29, 72, 72, 16) -4.591032
(29, 72, 72, 16) -5.843975
```

Run on a GeForce GTX 1080 Ti:

```
Python 3.6.3
Tensorflow 1.10.1
Numpy 1.15.1

2018-09-05 15:55:27.172809: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-09-05 15:55:27.685907: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties: 
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.582
pciBusID: 0000:8a:00.0
totalMemory: 10.91GiB freeMemory: 10.75GiB
2018-09-05 15:55:27.685952: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0
2018-09-05 15:55:28.161475: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-09-05 15:55:28.161524: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 
2018-09-05 15:55:28.161532: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N 
2018-09-05 15:55:28.161948: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10398 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:8a:00.0, compute capability: 6.1)

(29, 72, 72, 16) -4.591032
(29, 72, 72, 16) -5.843975
```

Run on a GeForce GTX 980 Ti:

```
Python 3.6.3
Tensorflow 1.5.1
Numpy 1.14.2

2018-09-05 16:05:54.557373: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-09-05 16:05:57.096399: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1105] Found device 0 with properties:
name: GeForce GTX 980 Ti major: 5 minor: 2 memoryClockRate(GHz): 1.076
pciBusID: 0000:02:00.0
totalMemory: 5.94GiB freeMemory: 5.83GiB
2018-09-05 16:05:57.096467: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1195] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 980 Ti, pci bus id: 0000:02:00.0, compute capability: 5.2)

(29, 72, 72, 16) -4.591032
(29, 72, 72, 16) -5.843975
```

Run on a K80:

```
Python 3.6.5
Tensorflow 1.5.1
Numpy 1.15.0

2018-09-05 14:37:31.279189: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-09-05 14:37:32.131417: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:895] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-09-05 14:37:32.131763: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1105] Found device 0 with properties:
name: Tesla K80 major: 3 minor: 7 memoryClockRate(GHz): 0.8235
pciBusID: 0000:00:1e.0
totalMemory: 11.17GiB freeMemory: 11.10GiB
2018-09-05 14:37:32.131795: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1195] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:00:1e.0, compute capability: 3.7)

(29, 72, 72, 16) -5.843975
(29, 72, 72, 16) -5.843975
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by gerbenvv** [comment]
Done. Please look at this as it has quite an impact on using 3D convolutions.

**Comment by gerbenvv** [comment]
This is not just a problem with large batch sizes, but rather, large inputs. The script:

```
import tensorflow as tf
import numpy as np



run_on_gpu = True  # Set to false for the correct result, computed on the CPU.



random_state = np.random.RandomState(42)

inputs = tf.placeholder(tf.float32, (None, 1, 1, 1, 1), 'inputs')

device_name = '/gpu:0' if run_on_gpu else '/cpu:0'

with tf.device(device_name):
    outputs = tf.tile(inputs, (1, 160, 320, 320, 16))
    weights = tf.constant(random_state.randn(1, 1, 1, 16, 16).astype(np.float32))
    outputs = tf.nn.conv3d(outputs, weights, strides=(1, 1, 1, 1, 1), padding='VALID')

session = tf.Session()

random_inputs = random_state.randn(1, 1, 1, 1, 1).astype(np.float32)

a = session.run(outputs, feed_dict={inputs: random_inputs})

print(a.shape, a.ravel()[-1])
```

Produces different results for the GTX 1080 Ti and CPU.

**Comment by gerbenvv** [comment]
@robieta Any progress?

**Comment by robieta** [comment]
Hi! Apologies for my late response.

I have replicated your finding that behavior on a 1080 is incorrect. Thank you so much for providing such a clear minimal reproduction.

I've opened an internal ticket with our op team, and I'll update this issue as that progresses.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drsxr** [comment]
That's really interesting.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gerbenvv** [comment]
@robieta It has been 2 and a half months. What is the status, because this bug has been confirmed and is impacting training and evaluating large 3D models.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by myron** [comment]
I can reproduce this too (TF1.12, CUDA9, Ubuntu 16, TitanXP 12GB and Tesla V100 16GB)
The 3D Conv result on GPU is completely wrong for certain input shapes/batch_sizes.

What triggers is:
1) having at least two 3D convolutions (with or without Relu in between).  a single 3D conv is OK
2) certain input shapes and batch sizes (some are OK, but many produce wrong result)
3) CPU result is always OK

I've implemented the same test cases in PyTorch and it always works correctly. 

Also downloading Nvidia docker container of Tensorflow works OK on GPU
https://www.nvidia.com/en-us/gpu-cloud/containers/
(those dockers Nvidia rebuilds from source for each Tensorflow version). 
   
So it seems the issues is specifically with "tensorflow-gpu" (pip built by Tensorflow)

It has been ~8 months this issue was reported.  It's unbelievable how long it takes to resolve it. (especially since this is a basic operation of any 3D network)

**Comment by gerbenvv** [comment]
I on the other hand, can no longer reproduce this bug with the snippets that I posted before with:

Tensorflow 1.31.1
CUDA 10.0
Nvidia driver 410.79
Ubuntu 16.04
Python 3.6.3
cudnn 7.4.2

on a GTX 1080 ti.

@myron Can you try upgrading to this setup? Do you have a simple reproduction snippet?

**Comment by myron** [comment]
Hi, so I can't test TF 1.13 yet (I cant upgrade to CUDA10 just yet). 
But for my setup, I get this bug (try both images_sizes)

TF 1.12.2
CUDA9
NVIDIA driver 418.56
Ubuntu 16.04
Python 3.6
cudnn 7.5

here is the snipper
```
import tensorflow as tf
import numpy as np

batch_size = 128  
image_size = (32,32,72,1)
#image_size = (64,72,64,1)

inputs = tf.placeholder(tf.float32, (None,)+image_size, 'inputs')
weights1 = tf.convert_to_tensor(np.random.random((3,3,3,1,32)), np.float32)
weights2 = tf.convert_to_tensor(np.random.random((1, 1, 1, 32, 16)), np.float32)
random_inputs = np.random.random((batch_size,)+image_size).astype(np.float32)

session = tf.Session()

with tf.device('/cpu:0'):
    outputs_cpu = tf.nn.conv3d(inputs, weights1, strides=(1,1,1,1,1), padding='SAME')
    outputs_cpu = tf.nn.conv3d(outputs_cpu, weights2, strides=(1,1,1,1,1), padding='SAME')
    a_cpu = session.run(outputs_cpu, feed_dict={inputs: random_inputs})

with tf.device('/gpu:0'):
    outputs_gpu = tf.nn.conv3d(inputs, weights1, strides=(1,1,1,1,1), padding='SAME')
    outputs_gpu = tf.nn.conv3d(outputs_gpu, weights2, strides=(1,1,1,1,1), padding='SAME')
    a_gpu = session.run(outputs_gpu, feed_dict={inputs: random_inputs})
  
print('CPUshape', a_cpu.shape, 'GPUshape', a_gpu.shape)
if np.allclose(a_cpu, a_gpu):
    print('GPU vs GPUa, OK (same)')
else:
    print('GPU vs GPUa, bad!!!')
```

**Comment by myron** [comment]
Alright, so upgrading to Tensorflow 1.31.1, CUDA 10.0 solved it for me. 
This issue can probably be closed now.
